noir_stdlib/src/hash.nr

```diff
@@ -74,3 +77,56 @@ pub fn poseidon2_permutation<N>(_input: [u8; N], _state_length: u32) -> [u8; N]
 
 #[foreign(sha256_compression)]
 pub fn sha256_compression(_input: [u32; 16], _state: [u32; 8]) -> [u32; 8] {}
+
+// Generic hashing support. 
+// Partially ported and impacted by rust.
+
+// Hash trait shall be implemented per type.
+trait Hash{
+    fn hash<H>(self, state: &mut H) where H: Hasher;
+
+    fn hash_slice<H>(data: [Self], state: &mut H) where H: Hasher{
+        for piece in data{
+            piece.hash(state);
+        }
+    }
+}
+
+// Hasher trait shall be implemented by algorithms to provide hash-agnostic means.
+trait Hasher{
+    fn finish(self) -> Field;
+    
+    fn write(&mut self, input: [Field]);
```

Comment:
We could probably make the types here generic. For example `sha` would want to write a list of bytes. This can most likely be left for a followup PR though. What do you think @jfecher ?

Reply:
I think fields are fine - this just wants some general integer that can represent other values and in that respect `Field` is the most general and doesn't imply range constraints.

Reply:
Forcing `Field` for the `HashMap` I think is fine, but the `Hasher` trait is more general than for just `HashMap` and can be applied to other algorithms/data structures. I think for this PR we can keep `Field` but just want to flag that this could be updated if a use-case ever arises where we do want the `Hasher` to use other types like bytes. For example, something that comes to mind is a merkle/verkle tree data structure that uses `keccak` to write circuits such as EVM storage proofs.

Reply:
I think making type generic here is indeed making sense; for now I left `TODO` comment (you guys can apply other means, standard for such situations).